File: 035.Search Insert Position.java
/*
    Given a sorted array and a target value, return the index if the target is found.
    If not, return the index where it would be if it were inserted in order.

    You may assume no duplicates in the array.

    Here are few examples.
    [1,3,5,6], 5 → 2
    [1,3,5,6], 2 → 1
    [1,3,5,6], 7 → 4
    [1,3,5,6], 0 → 0
     */

    public int searchInsert(int[] nums, int target) {
        if (nums.length == 0)
            return 0;

        return searchIndex(nums, target, 0, nums.length - 1);
    }

    public int searchIndex(int [] nums, int target, int begin, int end) {
        if (begin == end) {
            if (nums[begin] == target)
                return begin;
            if (target < nums[begin])
                return begin;
            if (target > nums[begin])
                return nums.length;
        }

        if (begin == end - 1) {
            if (nums[begin] >= target)
                return begin;
            if (nums[end] == target || (nums[begin] < target && nums[end] > target))
                return end;
            return nums.length;
        }

        int index = (begin + end) / 2;
        if (nums[index] == target)
            return index;

        return Math.min(searchIndex(nums, target, begin, index), searchIndex(nums, target, index + 1, end));
    }

File: 035. Search Insert Position.py
import math

class Solution(object):
    def searchInsert(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        start = 0
        end = len(nums) - 1

        ## 嘗試找出 target 在數列中的範圍
        while start + 1 < end:
            mid = int(start + (end - start)/2)

            if nums[mid] == target:
                return mid

            elif nums[mid] < target:
                start = mid

            elif nums[mid] > target:
                end = mid

        ## 測試 target 是否在數列中
        if nums[start] == target:
            return start

        if nums[end] == target:
            return end

        ## 確認 target 不存在數列中
        # target < 數列最左邊
        if target < nums[0]:
            return 0
        
        # target > 數列最右邊
        if target > nums[-1]:
            return len(nums)
        
        # target 在數列中間
        return end
    
